File: ios/Sources/MoQKit/Subscribe/internal/playback/JitterBuffer.swift
import Foundation

// MARK: - JitterBuffer

/// Generic sorted buffer with buffering/playing states for jitter resilience.
///
/// **Not thread-safe.** All access must be serialised by the owner (e.g. `VideoRendererTrack`).
///
/// `insert(item:timestampUs:)` returns `true` when the caller should fire a data-available
/// notification: either on the buffering→playing transition, or when inserting into an
/// empty buffer that is already playing.
final class JitterBuffer<T> {
    enum State {
        case buffering
        case pending
        case playing
    }

    struct Entry {
        let timestampUs: UInt64
        let item: T
    }

    private var targetBufferingUs: UInt64
    private var maxOffset: Int64?
    private var entries: [Entry] = []
    private var mode: State = .buffering
    private let wallClock: any PlaybackWallClock

    init(
        targetBufferingUs: UInt64,
        wallClock: any PlaybackWallClock = HostPlaybackWallClock()
    ) {
        self.targetBufferingUs = targetBufferingUs
        self.wallClock = wallClock
    }

    /// Insert an item sorted by timestamp. Returns `true` if the caller should fire
    /// a data-available notification (buffering→playing transition, or first item while playing).
    @discardableResult
    func insert(item: T, timestampUs: UInt64) -> Bool {
        guard let signedTimestamp = Self.signedTimestamp(timestampUs) else { return false }
        let result = signedTimestamp.subtractingReportingOverflow(wallClock.now(in: .us))
        guard !result.overflow else { return false }
        maxOffset = max(maxOffset ?? Int64.min, result.partialValue)

        let wasEmpty = entries.isEmpty
        let entry = Entry(timestampUs: timestampUs, item: item)

        let index = entries.firstIndex(where: { $0.timestampUs > timestampUs }) ?? entries.endIndex
        entries.insert(entry, at: index)

        // Transition buffering → playing when we have enough depth
        if mode == .buffering, updateBufferingStateIfReady() {
            return true
        }

        // Notify if inserting into empty buffer while playing
        return wasEmpty && mode == .playing
    }

    /// Dequeue the oldest entry. Returns `(nil, false)` when buffering or empty.
    /// Returns `(entry, false)` for frames that should be decoded but not displayed.
    func dequeue() -> (Entry?, Bool) {
        guard mode == .playing, !entries.isEmpty else { return (nil, false) }

        let entry = entries.removeFirst()

        let playable = targetPlaybackPTS().map { entry.timestampUs >= $0 } ?? true
        return (entry, playable)
    }

    /// Peek at the oldest entry without removing it. Returns nil when empty.
    func peekFront() -> Entry? {
        entries.first
    }

    /// Update the target buffering depth.
    @discardableResult
    func updateTargetBuffering(us: UInt64) -> Bool {
        targetBufferingUs = us
        guard mode == .buffering else { return false }
        return updateBufferingStateIfReady()
    }

    /// Clear all entries and reset to buffering state.
    func flush() {
        entries.removeAll()
        mode = .buffering
    }

    /// Returns the PTS of the first entry satisfying `predicate`, without removing it.
    func firstPts(where predicate: (Entry) -> Bool) -> UInt64? {
        entries.first(where: predicate)?.timestampUs
    }

    /// Unconditionally set the buffer state, bypassing normal transition logic.
    func setState(_ state: State) {
        mode = state
    }

    /// Remove the front entry unconditionally, ignoring `mode`.
    /// Returns `true` if an entry was removed.
    @discardableResult
    func discardFront() -> Bool {
        guard !entries.isEmpty else { return false }
        entries.removeFirst()
        return true
    }

    var state: State { mode }

    var count: Int { entries.count }

    var targetBuffering: UInt64 { targetBufferingUs }

    /// Estimated live-edge PTS from the maximum sender timestamp to local wall-clock offset.
    func estimatedLivePTS() -> Int64? {
        guard let maxOffset else { return nil }
        let result = wallClock.now(in: .us).addingReportingOverflow(maxOffset)
        guard !result.overflow else { return nil }
        return result.partialValue
    }

    /// Current desired playback PTS (`estimatedLivePTS - targetBuffering`).
    ///
    /// Returns `nil` when the result would be negative — callers (e.g. the video clock
    /// anchor in `VideoRenderer.startClockIfReady`) must fall back to the front-frame PTS
    /// instead of anchoring at 0, otherwise small-PTS streams would never display.
    func targetPlaybackPTS() -> UInt64? {
        guard let estimatedLivePTS = estimatedLivePTS() else { return nil }
        let bufferingSigned = Int64(clamping: targetBufferingUs)
        let result = estimatedLivePTS.subtractingReportingOverflow(bufferingSigned)
        guard !result.overflow, result.partialValue >= 0 else { return nil }
        return UInt64(result.partialValue)
    }

    /// PTS spacing between the first two buffered samples, when available.
    var frontFrameIntervalUs: UInt64? {
        guard entries.count >= 2 else { return nil }
        return entries[1].timestampUs > entries[0].timestampUs
            ? entries[1].timestampUs - entries[0].timestampUs
            : nil
    }

    /// Current buffered depth in milliseconds (newest − oldest entry timestamp).
    var depthMs: Double {
        guard entries.count >= 2 else { return 0 }
        return Double(entries.last!.timestampUs - entries.first!.timestampUs) / 1000.0
    }

    private func updateBufferingStateIfReady() -> Bool {
        guard mode == .buffering, entries.count >= 2 else { return false }
        let oldest = entries.first!.timestampUs
        let newest = entries.last!.timestampUs
        guard newest >= oldest, newest - oldest >= targetBufferingUs else { return false }
        mode = .playing
        return true
    }

    private static func signedTimestamp(_ timestamp: UInt64) -> Int64? {
        guard timestamp <= UInt64(Int64.max) else { return nil }
        return Int64(timestamp)
    }
}

File: android/moqkit/moqkit/src/main/java/com/swmansion/moqkit/subscribe/internal/playback/JitterBuffer.kt
package com.swmansion.moqkit.subscribe.internal.playback

import android.util.Log

private const val TAG = "JitterBuffer"

/**
 * Decode-order jitter buffer for media frames.
 *
 * Manages buffering->playing state transitions and playability decisions
 * based on wall-clock-to-PTS offset tracking.
 */
internal class JitterBuffer<T>(
    private var targetBufferingUs: Long,
    private val wallClockUs: () -> Long = { System.nanoTime() / 1000 },
) {
    enum class State { BUFFERING, PLAYING, PENDING }

    data class Entry<T>(
        val item: T,
        val timestampUs: Long,
        var offsetUs: Long,
    )

    private val lock = Object()
    private val entries = ArrayDeque<Entry<T>>()
    private var mode = State.BUFFERING
    private var maxOffset = Long.MIN_VALUE
    private var onDataAvailable: (() -> Unit)? = null
    var onStartPlaying: (() -> Unit)? = null
    var onStartBuffering: (() -> Unit)? = null

    private var exhausted = true

    fun setOnDataAvailable(callback: (() -> Unit)?) {
        synchronized(lock) { onDataAvailable = callback }
    }

    /** Force-set the buffer state, bypassing normal transition logic. */
    fun setState(state: State) {
        synchronized(lock) { mode = state }
    }

    /**
     * Peek at the front entry without removing it.
     * Returns the entry regardless of mode (works in PENDING, BUFFERING, and PLAYING).
     */
    fun peekFront(): Entry<T>? {
        synchronized(lock) { return entries.firstOrNull() }
    }

    /**
     * Peek at the first entry matching the predicate.
     */
    fun peekWhere(predicate: (Entry<T>) -> Boolean): Entry<T>? {
        synchronized(lock) { return entries.firstOrNull(predicate) }
    }

    /**
     * Unconditionally remove the front entry regardless of mode.
     * Returns true if an entry was removed.
     */
    fun discardFront(): Boolean {
        synchronized(lock) {
            if (entries.isEmpty()) return false
            entries.removeFirst()
            return true
        }
    }

    /**
     * Insert an item in decode order.
     * Tracks wall-clock-to-PTS offset and transitions buffering->playing.
     */
    fun insert(item: T, timestampUs: Long) {
        val notify: Boolean
        var startedPlaying = false
        synchronized(lock) {
            val offset = timestampUs - wallClockTimeUs()

            if (offset > maxOffset) {
                val diff = offset - maxOffset
                for (entry in entries) {
                    entry.offsetUs += diff
                }
                maxOffset = offset
            }

            val wasEmpty = entries.isEmpty()
            val entry = Entry(item, timestampUs, offset)

            // moq-mux emits frames in decode order. Presentation timestamps may move
            // backwards for B-frames, so sorting by PTS would corrupt decoder input order.
            entries.addLast(entry)

            // Transition buffering -> playing when we have enough depth
            if (mode == State.BUFFERING && entries.size >= 2) {
                val oldest = entries.minOf { it.timestampUs }
                val newest = entries.maxOf { it.timestampUs }
                if (newest - oldest >= targetBufferingUs) {
                    mode = State.PLAYING
                    Log.d(TAG, "Transitioned to PLAYING (${entries.size} frames buffered)")
                    notify = onDataAvailable != null
                    startedPlaying = true
                    return@synchronized
                }
            }

            // Notify if inserting into empty buffer while playing
            notify = exhausted || (wasEmpty && mode == State.PLAYING)
        }

        if (startedPlaying) onStartPlaying?.invoke()
        if (notify) {
            onDataAvailable?.invoke()
        }
    }

    /**
     * Dequeue the oldest entry.
     * Returns (entry, playable) where playable indicates the frame should be rendered vs dropped.
     * Returns (null, false) if buffering or empty.
     *
     * Playability is decided against the estimated video live-edge target
     * (`estimatedLivePTS - targetBufferingUs`), mirroring the iOS `dequeue()` path. The
     * finer-grained late-drop and render-scheduling decisions happen later in VideoRenderer
     * against the playback clock, so this gate intentionally carries headroom.
     */
    fun dequeue(): Pair<Entry<T>?, Boolean> {
        synchronized(lock) {
            if (mode != State.PLAYING || entries.isEmpty()) {
                if (mode != State.PENDING) exhausted = true
                return null to false
            }

            exhausted = false

            val entry = entries.removeFirst()
            val estimatedLivePts = wallClockTimeUs() + maxOffset
            val targetPlaybackPts = estimatedLivePts - targetBufferingUs
            val playable = entry.timestampUs >= targetPlaybackPts

            return entry to playable
        }
    }

    /**
     * Peek the PTS of the oldest buffered entry without removing it.
     * Returns null if not in PLAYING state or if buffer is empty.
     */
    fun peekNextTimestampUs(): Long? {
        synchronized(lock) {
            if (mode != State.PLAYING) return null
            return entries.firstOrNull()?.timestampUs
        }
    }

    /**
     * Estimated live-edge PTS from the maximum sender timestamp to local wall-clock offset.
     */
    fun estimatedLivePTS(): Long? {
        synchronized(lock) {
            if (maxOffset == Long.MIN_VALUE) return null
            return try {
                Math.addExact(wallClockTimeUs(), maxOffset)
            } catch (_: ArithmeticException) {
                null
            }
        }
    }

    /**
     * Current desired playback PTS (`estimatedLivePTS - targetBufferingUs`).
     */
    fun targetPlaybackPTS(): Long? {
        val estimated = estimatedLivePTS() ?: return null
        return try {
            Math.subtractExact(estimated, targetBufferingUs).takeIf { it >= 0L }
        } catch (_: ArithmeticException) {
            null
        }
    }

    /**
     * Estimated playback time in microseconds.
     * Used by VideoRenderer to compute render timestamps.
     */
    fun estimatedPlaybackTimeUs(): Long {
        synchronized(lock) {
            return wallClockTimeUs() + maxOffset - targetBufferingUs
        }
    }

    /**
     * Update target buffering depth.
     *
     * Returns true when a buffer that was waiting for enough depth became playable.
     */
    fun updateTargetBuffering(us: Long): Boolean {
        synchronized(lock) {
            targetBufferingUs = us
            return updateBufferingStateIfReady()
        }
    }

    /**
     * Discard all entries and reset to buffering state.
     */
    fun flush() {
        synchronized(lock) {
            entries.clear()
            mode = State.BUFFERING
            maxOffset = Long.MIN_VALUE
        }
        onStartBuffering?.invoke()
    }

    val state: State get() = synchronized(lock) { mode }
    val count: Int get() = synchronized(lock) { entries.size }
    val depthMs: Double get() = synchronized(lock) {
        if (entries.size < 2) 0.0
        else (entries.maxOf { it.timestampUs } - entries.minOf { it.timestampUs }).toDouble() / 1000.0
    }

    val frontFrameIntervalUs: Long? get() = synchronized(lock) {
        if (entries.size < 2) return@synchronized null
        val iterator = entries.iterator()
        val first = iterator.next().timestampUs
        val second = iterator.next().timestampUs
        if (second > first) second - first else null
    }

    private fun updateBufferingStateIfReady(): Boolean {
        if (mode != State.BUFFERING || entries.size < 2) return false
        val oldest = entries.minOf { it.timestampUs }
        val newest = entries.maxOf { it.timestampUs }
        if (newest < oldest || newest - oldest < targetBufferingUs) return false
        mode = State.PLAYING
        Log.d(TAG, "Transitioned to PLAYING (${entries.size} frames buffered)")
        return true
    }

    private fun wallClockTimeUs(): Long = wallClockUs()
}
